rust/actyx/ax-aql/src/language/parser.rs

```diff
@@ -721,6 +722,116 @@ pub(crate) fn query_from_str(s: &str) -> Result<Query<'_>> {
+        let mut errors = Vec::new();
+        query
+            .workflows
+            .iter()
+            .for_each(|(_ident, workflow)| errors.extend(self.check_workflow(workflow)));
+        errors
+    }
+
+    fn check_workflow(&'a self, workflow: &'a Workflow) -> Vec<QueryWorkflowAnalysisError> {
+        if self.workflow_tracker.contains(workflow) {
+            return Vec::new();
+        }
+        self.workflow_tracker.insert(workflow);
+        self.check_steps(&workflow.steps)
+    }
+
+    fn check_steps(&'a self, steps: &'a NonEmptyVec<WorkflowStep<'a>>) -> Vec<QueryWorkflowAnalysisError> {
+        let mut errors = Vec::new();
+        steps.iter().for_each(|step| errors.extend(self.check_step(step)));
+        errors
+    }
+
+    fn check_step(&'a self, step: &'a WorkflowStep<'a>) -> Vec<QueryWorkflowAnalysisError> {
+        use super::workflow::WorkflowStep::*;
+
+        match step {
+            Event { label, binders: _, .. } => {
+                let mut errors = Vec::new();
+                if !self.query.events.contains_key(&label) {
+                    errors.push(QueryWorkflowAnalysisError {
+                        error_type: QueryWorkflowAnalysisErrorType::UndeclaredEvent,
+                        span: label.clone(),
+                    });
+                }
+                // this part will also handle additional stuffs like type matching in `binders`
+                errors
+            }
+            Retry { steps } => self.check_steps(steps),
+            Timeout { steps, .. } => self.check_steps(steps),
+            Parallel { cases, .. } => {
+                let mut errors = Vec::new();
+                cases.iter().for_each(|case| errors.extend(self.check_steps(case)));
+                errors
+            }
+            Call {
+                workflow: workflow_ident,
+                cases: _,
+                ..
+            } => {
+                if let Some(workflow) = self.query.workflows.get(&*workflow_ident) {
+                    self.check_workflow(workflow)
+                } else {
+                    vec![QueryWorkflowAnalysisError {
+                        error_type: QueryWorkflowAnalysisErrorType::UndeclaredWorkflowCall,
+                        span: workflow_ident.clone(),
+                    }]
+                }
+            }
+            Compensate { body, with } => {
+                let mut errors = Vec::new();
+                errors.extend(self.check_steps(body));
+                errors.extend(self.check_steps(with));
+                errors
+            }
+            Choice { cases } => {
+                let mut errors = Vec::new();
+                cases.iter().for_each(|steps| errors.extend(self.check_steps(steps)));
+                errors
+            }
+        }
+    }
+}
+
+#[derive(Default)]
+pub(crate) struct WorkflowTracker<'a>(pub(crate) RefCell<BTreeSet<&'a Workflow<'a>>>);
```

Comment:
Probably no longer needed, but still: why `pub(crate)`? Such internal things should always stay as local as possible, especially with a RefCell inside.